### PR TITLE
Task app service deploy streaming log

### PR DESCRIPTION
**Pull request by ivywei0125**
What does this implement/fix? Explain your changes.
---------------------------------------------------
…

Does this close any currently open issues?
------------------------------------------

Any relevant logs, screenshots, error output, etc.?
-------------------------------------

Any other comments?
-------------------
…

Has this been tested?
---------------------------
- [ ] Tested

### CODE REVIEW

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/function/FunctionAppBase.java [hunk 1]
```diff
@@ -180,4 +178,12 @@ private boolean getIsRemoteDebuggingEnabled() {
                 StringUtils.equalsIgnoreCase(appSettings.get(JAVA_OPTS), getJavaOptsWithRemoteDebugEnabled(appSettings, appSettings.get(HTTP_PLATFORM_DEBUG_PORT)));
         return configEnabled && appSettingsEnabled;
     }
+
+    @Override
+    public boolean isLogStreamingEnabled() {
+        final OperatingSystem operatingSystem = Optional.ofNullable(getRuntime()).map(Runtime::getOperatingSystem).orElse(null);
+        final boolean isEnableApplicationLog = Optional.ofNullable(getDiagnosticConfig())
+                .map(DiagnosticConfig::isEnableApplicationLog).orElse(false);
+        return operatingSystem != OperatingSystem.LINUX && !isEnableApplicationLog;
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/function/FunctionAppBase.java]: reverted.

File: azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java [hunk 2]
```diff
@@ -210,7 +210,7 @@ private AppServiceConfig buildDefaultConfig(String subscriptionId, String resour
     private void deployArtifact(final FunctionAppBase<?, ?, ?> target) {
         final File file = new File(getDeploymentStagingDirectoryPath());
         final FunctionDeployType type = StringUtils.isEmpty(deploymentType) ? null : FunctionDeployType.fromString(deploymentType);
-        new DeployFunctionAppTask(target, file, type).doExecute();
+        new DeployFunctionAppTask(target, file, type, AzureMessager.getMessager(), true).doExecute();
```
Comment (wangmingliang-ms) [inline azure-functions-maven-plugin/src/main/java/com/microsoft/azure/maven/function/DeployMojo.java]: no need to pass messager as argument.

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AppServiceAppBase.java [hunk 3]
```diff
@@ -235,4 +235,8 @@ protected AppServiceKuduClient getKuduManager() {
         }
         return kuduManager;
     }
+
+    public boolean isEnableWebServerLogging() {
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/AppServiceAppBase.java]: `isStreamingLogSupported`

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/function/FunctionAppBase.java [hunk 4]
```diff
@@ -180,4 +178,12 @@ private boolean getIsRemoteDebuggingEnabled() {
                 StringUtils.equalsIgnoreCase(appSettings.get(JAVA_OPTS), getJavaOptsWithRemoteDebugEnabled(appSettings, appSettings.get(HTTP_PLATFORM_DEBUG_PORT)));
         return configEnabled && appSettingsEnabled;
     }
+
+    @Override
+    public boolean isEnableWebServerLogging() {
+        final OperatingSystem operatingSystem = Optional.ofNullable(getRuntime()).map(Runtime::getOperatingSystem).orElse(null);
+        final boolean isEnableApplicationLog = Optional.ofNullable(getDiagnosticConfig())
+                .map(DiagnosticConfig::isEnableApplicationLog).orElse(false);
+        return operatingSystem == OperatingSystem.LINUX || isEnableApplicationLog;
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/function/FunctionAppBase.java]: seems wrong, need to confirm. 
only windows function with application log enabled supports streaming log.

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/deploy/IOneDeploy.java [hunk 5]
```diff
@@ -32,7 +32,7 @@ default void deploy(DeployType deployType, File targetFile, String targetPath) {
 
     void deploy(DeployType deployType, File targetFile, DeployOptions deployOptions);
 
-    KuduDeploymentResult pushDeploy(DeployType var1, File var2, DeployOptions var3);
+    void pushDeploy(DeployType var1, File var2, DeployOptions var3);
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/deploy/IOneDeploy.java]: why modify signature?

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/task/DeployFunctionAppTask.java [hunk 6]
```diff
@@ -122,59 +111,31 @@ private File packageStagingDirectory() {
         }
     }
 
-    private void listHTTPTriggerUrls(FunctionApp target) {
+
+
+    private void startStreamingLog() {
+        if (!target.isEnableWebServerLogging() || !startStreamingLog) {
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/task/DeployFunctionAppTask.java]: `openStreamingLogOnFailure`

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/task/DeployFunctionAppTask.java [hunk 7]
```diff
@@ -122,59 +111,31 @@ private File packageStagingDirectory() {
...
+            return;
+        }
+        final OperatingSystem operatingSystem = Optional.ofNullable(target.getRuntime()).map(Runtime::getOperatingSystem).orElse(null);
+        if (operatingSystem == OperatingSystem.LINUX) {
+            return;
+        }
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/task/DeployFunctionAppTask.java:119]: no need to check again.

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/task/DeployFunctionAppTask.java [hunk 8]
```diff
@@ -122,59 +111,31 @@ private File packageStagingDirectory() {
...
-        final int[] count = {0};
-        return Mono.fromCallable(() -> {
-                final AzureString message = count[0]++ == 0 ? AzureString.fromString(LIST_TRIGGERS) : AzureString.format(LIST_TRIGGERS_WITH_RETRY, count[0], LIST_TRIGGERS_MAX_RETRY);
-                messager.info(message);
-                return Optional.of(functionApp.listFunctions())
-                    .filter(CollectionUtils::isNotEmpty)
-                    .orElseThrow(() -> new AzureToolkitRuntimeException(NO_TRIGGERS_FOUNDED));
-            }).subscribeOn(Schedulers.boundedElastic())
-            .retryWhen(Retry.fixedDelay(LIST_TRIGGERS_MAX_RETRY - 1, Duration.ofSeconds(LIST_TRIGGERS_RETRY_PERIOD_IN_SECONDS))).block();
+    private void stopStreamingLog() {
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/task/DeployFunctionAppTask.java]: synchronized.

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/task/DeployFunctionAppTask.java [hunk 9]
```diff
@@ -122,59 +111,31 @@ private File packageStagingDirectory() {
...
-    private List<FunctionEntity> listFunctions(final FunctionApp functionApp) {
-        final int[] count = {0};
-        return Mono.fromCallable(() -> {
-                final AzureString message = count[0]++ == 0 ? AzureString.fromString(LIST_TRIGGERS) : AzureString.format(LIST_TRIGGERS_WITH_RETRY, count[0], LIST_TRIGGERS_MAX_RETRY);
-                messager.info(message);
-                return Optional.of(functionApp.listFunctions())
-                    .filter(CollectionUtils::isNotEmpty)
-                    .orElseThrow(() -> new AzureToolkitRuntimeException(NO_TRIGGERS_FOUNDED));
-            }).subscribeOn(Schedulers.boundedElastic())
-            .retryWhen(Retry.fixedDelay(LIST_TRIGGERS_MAX_RETRY - 1, Duration.ofSeconds(LIST_TRIGGERS_RETRY_PERIOD_IN_SECONDS))).block();
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/task/DeployFunctionAppTask.java:178]: move them to `FunctionApp` instead.

File: azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/task/DeployWebAppTask.java [hunk 10]
```diff
@@ -46,8 +36,10 @@ public class DeployWebAppTask extends AzureTask<WebAppBase<?, ?, ?>> {
     private final WebAppBase<?, ?, ?> webApp;
     private final List<WebAppArtifact> artifacts;
     private final boolean restartSite;
+    private final boolean startStreamingLog;
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-appservice-lib/src/main/java/com/microsoft/azure/toolkit/lib/appservice/task/DeployWebAppTask.java]: `openStreamingLogOnFailure`.

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/task/DeploySpringCloudAppTask.java [hunk 11]
```diff
@@ -104,7 +106,9 @@ private List<AzureTask<?>> initTasks() {
             draft.updateIfExist();
         }));
         tasks.add(new AzureTask<Void>(app::reset));
-        tasks.add(new AzureTask<Void>(this::stopStreamingLog));
+        if (this.waitDeploymentComplete) {
+            tasks.add(new AzureTask<Void>(this::startAppService));
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/task/DeploySpringCloudAppTask.java]: `startApp`

File: azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/task/DeploySpringCloudAppTask.java [hunk 12]
```diff
@@ -127,19 +131,19 @@ private void startStreamingLog(boolean follow) {
         final String instanceName = instanceList.stream().max(Comparator.comparing(DeploymentInstance::startTime))
                 .map(DeploymentInstance::name).orElse(null);
         Optional.ofNullable(instanceName).ifPresent(i -> {
-            messager.info(AzureString.format("Starting app({0}) and opening streaming log of instance({1})...", this.config.getAppName(), instanceName));
+            messager.info(AzureString.format("Opening streaming log of instance({0})...", instanceName));
             messager.debug("###############STREAMING LOG BEGIN##################");
-            this.streamingLogDisposable = this.deployment.streamLogs(i, 300, 500, 1024 * 1024, follow)
+            this.deployment.streamLogs(i, 300, 500, 1024 * 1024, follow)
```
Comment (wangmingliang-ms) [inline azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/task/DeploySpringCloudAppTask.java:136]: where are these numbers from?

Comment (ivywei0125) [reply azure-toolkit-libs/azure-toolkit-springcloud-lib/src/main/java/com/microsoft/azure/toolkit/lib/springcloud/task/DeploySpringCloudAppTask.java:136]: https://github.com/Azure/azure-cli-extensions/blob/main/src/spring/azext_spring/app.py#L410

**Review by wangmingliang-ms — DISMISSED** [review]
LGTM

**Review by Flanker32 — APPROVED** [review]
LGTM

**Pull request merged**